### PR TITLE
CM-TCK use osgi-test support

### PR DESCRIPTION
**Pull request by stbischof**
I am using Ubuntu. TCK did not run  befor usage of `osgi-tests` now it runs locally without issues

- Testing wheter this runs on  ci
- what would be the best way to add `osgi-test.junit5.cm` bundle  and blowing all other test away.

### CODE REVIEW

File: org.osgi.test.cases.framework/src/org/osgi/test/cases/framework/junit/context/InstallBundleTest.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
+/*******************************************************************************
```
Comment (tjwatson) [inline org.osgi.test.cases.framework/src/org/osgi/test/cases/framework/junit/context/InstallBundleTest.java:1]: What does this have to do with the CM TCK?

File: org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/ConfigurationPluginTests.java [hunk 2]
```diff
@@ -87,21 +115,25 @@ protected void setUp() throws Exception {
         }
 	}
 
+	private long getLongProperty(String string, int i) {
```
Comment (tjwatson) [inline org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/ConfigurationPluginTests.java:118]: What is this used for?

File: org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/CMCoordinationTestCase.java [hunk 3]
```diff
@@ -345,4 +372,8 @@ public void updated(Dictionary<String, ? > properties)
 		assertFalse(events.get(3));
 
 	}
+
```
Comment (tjwatson) [inline org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/CMCoordinationTestCase.java:375]: Why do we have this?

File: org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/ConfigurationPluginTests.java [hunk 4]
```diff
@@ -250,14 +292,15 @@ public void testRankingForManagedService() throws Exception {
...
-		cm.getConfiguration(pid).delete();
+		cm.waitForService(500).getConfiguration(pid).delete();
 		trace("Wait until the ManagedService has gotten the delete");
 
 		assertTrue("Update done",
 				synchronizer.waitForSignal(SIGNAL_WAITING_TIME));
 		assertNull(synchronizer.getProps());
 		assertTrue(context.invocationOrder.isEmpty());
 		assertTrue(context.errors.isEmpty());
+		reg.unregister();
```
Comment (tjwatson) [inline org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/ConfigurationPluginTests.java:303]: This service remaining in the registry may be the root cause of the issues

File: org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/ConfigurationPluginTests.java [hunk 5]
```diff
@@ -335,15 +380,19 @@ public void testRankingForManagedServiceFactory() throws Exception {
...
```
Comment (tjwatson) [inline org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/ConfigurationPluginTests.java:383]: This service in the registry may be the root cause of the issues.

File: org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/ConfigurationPluginTests.java [hunk 6]
```diff
@@ -362,15 +412,18 @@ public void testTargettingManagedService() throws Exception {
 		assertEquals("val", props.get("key"));
 		verifyPlugins(props, false, Arrays.asList(1, 2), Arrays.asList(1, 2),
 				context);
+		reg.unregister();
```
Comment (tjwatson) [inline org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/ConfigurationPluginTests.java:415]: Same comment, good find!

File: org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/ConfigurationPluginTests.java [hunk 7]
```diff
@@ -390,15 +444,17 @@ public void testTargettingManagedServiceFactory() throws Exception {
 		assertEquals("val", props.get("key"));
 		verifyPlugins(props, true, Arrays.asList(1, 2), Arrays.asList(1, 2),
 				context);
+		reg.unregister();
```
Comment (tjwatson) [inline org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/ConfigurationPluginTests.java:447]: +1

**Review by tjwatson — COMMENTED** [review]
Overall this looks like a really good direction.  I think we should consider this for an R8 fix.

File: org.osgi.test.cases.framework/src/org/osgi/test/cases/framework/junit/context/InstallBundleTest.java (hunk 1 above)
Comment (stbischof) [reply org.osgi.test.cases.framework/src/org/osgi/test/cases/framework/junit/context/InstallBundleTest.java:1]: Nothing.

I just had some errors and had to find out whether this ins an tck Design error or something that came from Framework...

File: org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/ConfigurationPluginTests.java (hunk 2 above)
Comment (stbischof) [reply org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/ConfigurationPluginTests.java]: The original test hat optional to set the sleep time.. this is just something i will clean up when tests work every time

File: org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/ConfigurationPluginTests.java (hunk 4 above)
Comment (stbischof) [reply org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/ConfigurationPluginTests.java]: Yes.  The unregisting of the managed services is an issue. I expected that the closeable bundlecontext of osgi-tests will unregister this. Bur this did not happed. Need more investigation here

File: org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/CMCoordinationTestCase.java (hunk 3 above)
Comment (stbischof) [reply org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/CMCoordinationTestCase.java]: sorry i reset the commit. so i do not know what happened here

File: org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/CMControl.java [hunk 8]
```diff
@@ -2294,10 +2362,10 @@ public void updated(final Dictionary<String, ?> properties) throws Configuration
 			assertNull(updates.get(0));
 			assertEquals("somevalue", updates.get(1).get("scalar"));
 			assertEquals("newvalue", updates.get(2).get("scalar"));
-		} finally {
-			this.unregisterService(listener);
-			this.unregisterService(ms);
-		}
+			reg.unregister();
```
Comment (stbischof) [inline org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/CMControl.java:2365]: same here @tjwatson

File: cnf/repo/org.osgi.impl.service.cm/org.osgi.impl.service.cm-7.0.0.lib [hunk 9]
```diff
@@ -1 +1,2 @@
 org.apache.felix.configadmin; version=1.9
+org.osgi.test.junit5.cm; version=latest
```
Comment (bjhargrave) [inline cnf/repo/org.osgi.impl.service.cm/org.osgi.impl.service.cm-7.0.0.lib:2]: A testing jar must not be in the impl lib file!

File: org.osgi.test.cases.cm/bnd.bnd [hunk 10]
```diff
@@ -30,19 +30,23 @@ Import-Package: ${-signaturetest}, *
...
+-buildpath: \
+	org.osgi.test.support;version=project,\
+	org.osgi.framework;maven-scope=provided;version='1.8',\
+	org.osgi.resource;maven-scope=provided;version='1.0',\
+	org.osgi.service.permissionadmin;maven-scope=provided;version='1.2',\
+	org.osgi.namespace.implementation;version='1.0',\
+	org.osgi.namespace.service;version='1.0',\
+	org.osgi.service.coordinator;version='1.0',\
+	org.osgi.service.cm;version=latest,\
+	org.osgi.annotation.versioning,\
```
Comment (bjhargrave) [inline org.osgi.test.cases.cm/bnd.bnd:42]: Why does the TCK need versioning annotations? And it does need them, it should use `osgi.annotations;version=8.0`.

File: org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/CMControl.java [hunk 11]
```diff
@@ -217,29 +254,43 @@
 
 	private static final String neverlandLocation = "http://neverneverland/";
 
-	protected void setUp() throws Exception {
+	@BeforeEach
+	void setUp(TestInfo testInfo) throws Exception {
+		name = testInfo.getTestMethod()
+				.map(Method::getName)
+				.orElse("AutoName_" + System.currentTimeMillis());
```
Comment (bjhargrave) [inline org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/CMControl.java:261]: use `orElseGet` to avoid the string building when unnecessary.

File: org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/CMControl.java [hunk 12]
```diff
@@ -311,17 +361,19 @@ private void printoutPermissions() {
 	}
 
 	private void setBundlePermission(Bundle b, List<PermissionInfo> list) {
-		if (permAdmin == null)
+		if (permAdminA.getService() == null)
```
Comment (bjhargrave) [inline org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/CMControl.java:364]: You should put `permAdminA.getService()` into a variable and then test for `null` and use. Calling `permAdminA.getService()` twice could give different results.

File: org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/CMControl.java [hunk 13]
```diff
@@ -2294,10 +2367,10 @@ public void updated(final Dictionary<String, ?> properties) throws Configuration
 			assertNull(updates.get(0));
 			assertEquals("somevalue", updates.get(1).get("scalar"));
 			assertEquals("newvalue", updates.get(2).get("scalar"));
-		} finally {
-			this.unregisterService(listener);
-			this.unregisterService(ms);
-		}
+			reg.unregister();
+			// TODO BJ:WHY MUST THIS BE DONE MANUALLY????
```
Comment (bjhargrave) [inline org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/CMControl.java:2371]: I assume that since the test method uses the raw service API to register a service, it must then unregister it as no one else will.

**Review by bjhargrave — COMMENTED** [review]
Wow! There is a lot of nice work in here. @tjwatson Do you think this is worth merging at this point in the release?

File: cnf/repo/org.osgi.impl.service.cm/org.osgi.impl.service.cm-7.0.0.lib (hunk 9 above)
Comment (stbischof) [reply cnf/repo/org.osgi.impl.service.cm/org.osgi.impl.service.cm-7.0.0.lib:2]: sure but if i put it into the junit.tck all other test will fail because of the  missing cm and adding it just to the bndrun file just works in ide but not on ci because it does not find it on one of the given repos.

Comment (bjhargrave) [reply cnf/repo/org.osgi.impl.service.cm/org.osgi.impl.service.cm-7.0.0.lib:2]: There is no reason some other project using the CM impl needs this osgi-test jar. You already have it in the -runbundles and -buildpath of the TCK which should be sufficient. Yes, you cannot put it in the osgi.tck.junit-platform.bnd file.

File: org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/CMCoordinationTestCase.java (hunk 3 above)
Comment (tjwatson) [reply org.osgi.test.cases.cm/src/org/osgi/test/cases/cm/junit/CMCoordinationTestCase.java:380]: I was asking about the getLongProperty method you had added as a stub method, but it is gone now.

**Comment by tjwatson** [comment]
> Wow! There is a lot of nice work in here. @tjwatson Do you think this is worth merging at this point in the release?

Only if we have some proof that it makes the TCK pass more reliably.  But so far I have not seen that.  I really think we may be chasing a bug in Felix CM implementation at this point.
